### PR TITLE
Review request - attempt at Mojo::DynamicMethods

### PR DESCRIPTION
**Pull request by shadowcat-mst**
### Summary

I created a thing called Mojo::DynamicMethods and use it to kill AUTOLOAD

### Motivation

sri nerd sniped me by telling me to kill AUTOLOAD

### References

I hate you all

### CODE REVIEW

**Comment by kraih** [comment]
String evals and the tricky garbage collection code have me a little worried. The performance gains are amazing, but i don't think it's worth the cost in maintainability. Especially when compared to your other solution, which works almost as well.
```diff
diff --git a/lib/Mojolicious/Controller.pm b/lib/Mojolicious/Controller.pm
index dcd200c74..abee5da5b 100644
--- a/lib/Mojolicious/Controller.pm
+++ b/lib/Mojolicious/Controller.pm
@@ -19,6 +19,9 @@ my %RESERVED = map { $_ => 1 } (
   qw(namespace path status template text variant)
 );

+{package Mojolicious::Controller::_Helpers};
+unshift our @ISA, 'Mojolicious::Controller::_Helpers';
+
 sub AUTOLOAD {
   my $self = shift;

@@ -29,9 +32,26 @@ sub AUTOLOAD {
   # Call helper with current controller
   Carp::croak qq{Can't locate object method "$method" via package "$package"}
     unless my $helper = $self->app->renderer->get_helper($method);
+
+  Mojo::Util::monkey_patch 'Mojolicious::Controller::_Helpers', $method, sub {
+    my $self = shift;
+    Carp::croak
+      qq{Can't locate object method "$method" via package "${\ref $self}"}
+      unless my $helper = $self->{helpers}{$method}
+      ||= $self->app->renderer->get_helper($method);
+    return $self->$helper(@_);
+  };
+
   return $self->$helper(@_);
 }

+sub can {
+  my ($self, $method, @rest) = @_;
+  return undef unless my $can = $self->SUPER::can($method, @rest);
+  my $h = ${Mojolicious::Controller::_Helpers}{$method}{CODE};
+  return $h && $h eq $can ? undef : $can;
+}
+
 sub continue { $_[0]->app->routes->continue($_[0]) }

 sub cookie {
```

**Comment by kraih** [comment]
Actually, i think we can avoid most of the downsides of the other solution by moving the cache to the app instance.
```diff
diff --git a/lib/Mojolicious/Controller.pm b/lib/Mojolicious/Controller.pm
index dcd200c74..149992cb2 100644
--- a/lib/Mojolicious/Controller.pm
+++ b/lib/Mojolicious/Controller.pm
@@ -19,6 +19,9 @@ my %RESERVED = map { $_ => 1 } (
   qw(namespace path status template text variant)
 );

+{package Mojolicious::Controller::_Helpers};
+unshift our @ISA, 'Mojolicious::Controller::_Helpers';
+
 sub AUTOLOAD {
   my $self = shift;

@@ -29,9 +32,27 @@ sub AUTOLOAD {
   # Call helper with current controller
   Carp::croak qq{Can't locate object method "$method" via package "$package"}
     unless my $helper = $self->app->renderer->get_helper($method);
+
+  Mojo::Util::monkey_patch 'Mojolicious::Controller::_Helpers', $method, sub {
+    my $self = shift;
+    my $app  = $self->app;
+    Carp::croak
+      qq{Can't locate object method "$method" via package "${\ref $self}"}
+      unless my $helper = $app->{helpers}{$method}
+      ||= $app->renderer->get_helper($method);
+    return $self->$helper(@_);
+  };
+
   return $self->$helper(@_);
 }

+sub can {
+  my ($self, $method, @rest) = @_;
+  return undef unless my $can = $self->SUPER::can($method, @rest);
+  my $h = ${Mojolicious::Controller::_Helpers}{$method}{CODE};
+  return $h && $h eq $can ? undef : $can;
+}
+
 sub continue { $_[0]->app->routes->continue($_[0]) }

 sub cookie {
```

**Comment by shadowcat-mst** [comment]
If "kill AUTOLOAD entirely" is no longer the goal, my first attempt was definitely simpler, and while "moving the cache to the app" is only going to work for controllers, that's probably the thing that most needs optimising anyway.

Though I'd note the GC thing is fairly idiomatic in stuff that uses weak references this way (it's essentially an unrolling of Hash::Util::FieldHash's idhash stuff that moves the effort somewhere that isn't as hot a path), and I'd be happy to comment it up - and the string eval could be got rid of by basically copy-pasting the croak and having the builder coderef passed to DynamicMethods->import.

So I *could* achieve "less crack, full implementation" - but in terms of *performance* so far as users will notice it, sri's adjusted version of my first attempt, plus perhaps duplicating that approach into Validation since that's also per-request calls, would work out fine.

Which I think leaves my considered opinion as "I still basically like both my patches, each for the purpose it was implemented for, and it's mostly a question of which purpose the Mojo team considers more important". Happy to work on tweaked versions of either if people want.

**Comment by shadowcat-mst** [comment]
Branch updated to strip the string eval

File: lib/Mojo/DynamicMethods.pm [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  }
+}
+
+sub register_dynamic {
+  my ($target, $object, $name, $code) = @_;
+  state %dyn_methods;
+  {
+    state %done;
+    unless ($done{$target}{$name}) {
+      our %Builders;
```
Comment (kraih) [inline lib/Mojo/DynamicMethods.pm]: You forgot one `our %Builders`.

**Comment by shadowcat-mst** [comment]
Took sri's mojofication and then moved registration to support direct ->add_helper because users.

**Comment by shadowcat-mst** [comment]
Hokay, I think that's unit tests and POD at least ish done. Comments welcome.

File: lib/Mojo/DynamicMethods.pm [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+Registers the method C<$name> as eligible for dynamic dispatch for C<$class>,
+and sets C<$code> to be looked up for C<$name> by reference C<$ref> in a
+dynamic method contructed by L</BUILD_DYNAMIC>.
+
+=head1 METHODS
+
+L<Mojo::DynamicMethods> requires you to implement the following method on
+any class participating in dynamic dispatch:
+
+=head2 BUILD_DYNAMIC
```
Comment (kraih) [inline lib/Mojo/DynamicMethods.pm]: We only document methods that the class actually contains like this.

**Comment by kraih** [comment]
`Mojolicious::Validator::new` needs documentation added, like the other [overloaded constructors](https://github.com/mojolicious/mojo/blob/master/lib/Mojo/Log.pm#L241) we have. And i guess the `BUILD_DYNAMIC` methods need to be hidden from `Test::Pod::Coverage`.

**Comment by kraih** [comment]
Please review and vote @jberger, @marcusramberg and @jhthorsen.

**Comment by shadowcat-mst** [comment]
I converted the AUTOLOAD sections to BUILD_DYNAMIC sections. Not sure if that's quite right but it seemed like a good starting point.

**Comment by shadowcat-mst** [comment]
My gods, it's full of green.

Apologies for forgetting how to POD.

**Comment by marcusramberg** [comment]
I think this looks great, and am 👍 to this change.

**Comment by shadowcat-mst** [comment]
@kraih I kept it as capitalised because it's still morally a global but lowercase does look better so I've changed it, thanks.

I definitely think ::_Dynamic looks better than ::_DynamicMethods did in stack traces etc. but I'm about out of opinions at this point

**Comment by shadowcat-mst** [comment]
@kraih asked about possible downsides, including "new leaks".

So, given I'm using the fieldhash trick to make coderefs go out of scope as soon as the object they're eligible to be called on does, I don't *think* I've created any new leaks - the only leak I can think of is where one of those coderefs closes over the thing they're a helper for, which while unfortunate would already be a leak on master AFAICT.

The other thing this code will do is blow the perl5 VM method cache every time it registers a new method name, but in any remotely long lived process that shouldn't make any real difference - if you registered a gensym()ed method name on every request then *maybe* it would be detectable but I really wouldn't expect it to be significant on its own, let alone compared to the performance gains already inherent in this patchset.

I guess a theoretical disadvantage is that if somebody subclassed one of the -dispatch classes and then implemented their own AUTOLOAD, they'd no longer have that called in the case of an already-registered dynamic method name, but that's into jenga style crack stacking so I'm not convinced we should care.

**Comment by kraih** [comment]
The code does look pretty clean now, and i don't see any obvious risks or downsides anymore, so i vote 👍 too.

**Pull request merged**